blog/_posts/2017-12-05-ml&pl.md

```diff
@@ -0,0 +1,88 @@
+
+Consider a world where adding new hardware support – or new kinds of data representations – could easily be accomplished by a user in high-level code, without changes to the original system. We expect ML systems to take inspiration from existing numerical computing languages, which can [already handle these tasks](https://arxiv.org/pdf/1604.03410.pdf) with ease.
+
+Type systems can also have safety benefits, but current ones are not suited to array-heavy code where array dimensions are meaningful (for example, spatial vs channel vs batch dimensions in images). These distinctions are left to [pure convention](https://github.com/pytorch/pytorch/issues/1220), and hairy dimension-permuting code has no protection from mistakes, leaving much room for more array-aware type systems. We expect the trend towards dynamic typing to continue[^types], mainly due to practitioners’ preference for interactivity and scripting, but hope to see further innovations like [CNTK’s optionally dynamic dimensions](https://cntk.ai/pythondocs/sequence.html).
+
+[^types]: Though we note that, internally, current systems span the gamut from fully dynamic (PyTorch and its ATen backend) to unusually static (TensorFlow’s XLA and MXNet, where all dimensions are known before the graph is run).
+
+ML engineers are increasingly interested in traditional [software engineering problems](https://papers.nips.cc/paper/5656-hidden-technical-debt-in-machine-learning-systems.pdf) like the maintenance and extension of production systems. The [ML programming model](https://medium.com/@karpathy/software-2-0-a64152b37c35) makes it harder to create abstraction barriers and interfaces between components, and re-training of a model can easily break backwards compatibility. ML languages will likely be able to incorporate solutions to these problems just as regular languages do.
+
+A downside to any new language is that it require a new library ecosystem, as only code written for the new runtime benefits from it. For example, rather than reusing the Python ecosystem, the TensorFlow developers need to rewrite libraries for tasks like [image processing](https://www.tensorflow.org/api_guides/python/image) and [file IO](https://www.tensorflow.org/api_docs/python/tf/TextLineReader) in the graph language, throwing out the vast effort behind projects like SciPy. This may well be the only way forward, but ML practitioners should not split from the wider numerical and HPC community. An ideal ML ecosystem is an ideal numerical one, and vice versa, and collaboration between these communities will multiply everyone’s efforts.
```

Comment:
Mention that this is something that Julia tools are able to do well with

Reply:
Note that our goal with this is not to evangelise Julia specifically, but just to outline the trend in ML (which may have other end points, like TF becoming a full language). In future posts we'll discuss Julia's tools in more detail and link them back to these ideas explicitly.

Reply:
Well usually it's nice to add to "an ideal ML ecosystem" would ..., by mentioning some recent efforts to solve the problem. The Julia tools do this, so they should be mentioned. I'm sure there's some Nim, Rust, Python, etc. tool out there that's trying to do the same that can also get a mention, but I don't know them.